### DISCUSSION TITLE
Report - (4th Sem.) Week 6📢

### DISCUSSION THREAD

**Discussion by giomskii** [SE-TINF22B6/DHBWhub#681] [Weekly Report]
Greetings and welcome to our fourteenth progress report! We're thrilled to update you on the latest news and developments since last week! 🥳

### Metrics Metrics Metrics 📊 
Let's start with a quick summary of our topic - Software testing metrics. You can also find our slides right [here](https://github.com/SE-TINF22B6/DHBWhub/blob/master/docs/SE_Metrics.pptx).
  
#### Software Testing Metrics 🧪 
In general, software testing metrics are quantifiable indicators that represent the overall status, efficiency and quality of not only the software tests regarding the product, but also the testing process itself. These metrics are important for identifying serious problems beforehand, optimizing the allocation of resources that go into testing and finally, the monitoring of the process development leading to the measurement of the continous improvement of the testing procedure.  
  
The metrics can roughly be divided into three categories:
1. Process Metrics - KPIs regarding the status of the processes that are part of the software developments life cycle.
2. Product Metrics - KPIs that are solely describing the testing quality of the software product.
3. Project Metrics - KPIs depicting the overall quality of the project itself.
  
It is also important to mention that not every KPI is necesseraly important for the regarded software project. In order to find out whether an metric is needed or not, you have to go by the following test metrics life cycle:  
1. Analysis: Identifiying the best metrics for testing purposes.
2. Communication: Creating common ground between developer team and stakeholder.
3. Evaluation: Execution of test metrics.
4. Reporting: Summarization of test metrics and report to stakeholders for better improvement.
  
The following bullet points contain some metric examples in regards to software testing metrics.:
 - Defect Density: Amount of errors and mistakes / Totol amount of lines of code.
 - Test Coverage: Amount of lines covered by test cases / Total amount of lines of code.
 - Time To Repair Defects: Timestamp where defect is fixed - Timestamp where defect was reported.
 - Execution rate of test cases: Actually executed test cases / Total of planned test cases.
  
#### Three additional metrics ❓ ❔ ❓ 
Currently, we have agreed to measure these three metrics in addition to the software testing metrics:
1. Coupling:  
As modularity is one part of our non-functional requirements analysis, it is important to check that our frontend components have loose coupling since each component should be reusable as easy as possible. Hereby, we can use the formulas of Fenton and Melton which for example include efferent 'CE' and afferent coupling 'CA', instability 'I' and abstractness 'A'. You can find the specific formula [here](https://www.techtarget.com/searchapparchitecture/tip/The-basics-of-software-coupling-metrics-and-concepts).
2. Web Application Performance:  
The performance of our webpage can be measured by many tools, such as 'PageSpeed Insights' for metrics regarding the amount of time needed for rendering our webpage, or the 'Firefox Performance Monitor', which gives us insights on how each user interaction affects the websites overall performance (e.g. delay time).
3. Web Application Monitoring:  
It goes without saying that most of the user experience and interaction happens on the webpage itself. A fast response time in particular is an important non-functional criteria to keep our users in a good mood. This can be achieved by tools such as 'Pingdom' or 'Google Lighthouse'. These analyze our website in several different categories and return a score for each one.

### Developments on the Frontend 💻 
The frontend is now able to receive data from the backend! We have also restricted unsigned users from interacting with posts and other features only signed in members are able to do. Moreover, we have implemented the frontend-facade for signing in via Google-login. The button in the picture below redirects you to Google, where you are able to select the account you want to sign in with:  
  
![Google-Sign-In](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/423938e5-61d3-43a7-8bd1-2ed2c5bb9465)
  
We have also worked on our final mockups, such as the Profile, Imprint, Terms of Service and Privacy Policy pages which you can see below:  
  
![grafik](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/d97f27d3-b1b8-4059-9086-937ff72a663a)

![grafik](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/b4290e3e-9d0b-4463-a92c-1745d428e886)

![grafik](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/307953d7-1b83-42ea-8941-cc099c945751)

![grafik](https://github.com/SE-TINF22B6/DHBWhub/assets/122597204/07082b04-5852-49cd-a5c1-66c36d14d6a1)
  
### Backend and Database 📚 
We have implemented a notification system which tracks every user activity and reports it to the users who are affected by it. Not only that, but several actions on the same component accumulate to a single notification! We have done something similary with tracking like activities, since in every login, the web page has to know which posts, comments and events are liked by the user. In the next weeks, we hope to show you this function on the actual website!

  
### Sprint 9: What have we achieved ❓ What's to come❔
✅ Frontend: Setting up Google OAuth Authorization buttons
✅ Frontend: Post component retrieves post data from backend
✅ Frontend: Adapt event component to retrieve data from backend
✅ Frontend: Adapt components to retrieve profile picture from backend
✅ Frontend: Implement tests for essential homepage logic
✅ Postman: Updated Postman Collection with new Request Methods
✅ Backend: Implemented concept for logging like activities
✅ Backend: Implemented notifications concept
✅ Backend: Retrieve most popular tags 
✅ Backend: Adjusted SavedPost Endpoints
🚧 Frontend: Adapting Sign-Up dialog
🚧 Frontend: Implement profile page
🚧 Frontend: Implement 'Create Post' popup window
🚧 Frontend: Implement notifications window
🚧 Frontend: Implement comment-component
🚧 Frontend: Implement FAQ-page
🚧 Frontend: Implement friends page
🚧 Frontend: Implement privacy policy page
🚧 Frontend: Implement calendar page
🚧 Frontend: Enable registration via google
🚧 Frontend: Implement 'Forget password dialog'
🚧 Backend: Setup HTTPS for backend server
🚧 Backend: Implement report function
🚧 Backend: Refine Post Endpoints
🚧 Backend: Implement Endpoint for posts of a single user
🚧 Backend: Insert all Course options

**Comment by BenjaminB109** [comment]
Dear DHBWhub Team,
I think the monitoring for the topics Coupling, Web Application Performance and Web Application Monitoring are well selected for your project. Keep up the good work.

**Comment by fehlend0** [comment]
Hello DHBWHub Team, 

Your comprehensive breakdown of software testing metrics provides a clear roadmap for ensuring top-notch quality and efficiency. The sprint achievements are impressive, from OAuth Authorization buttons to notifications concepts. 

Looking forward to seeing the upcoming implementations,
Your Statify Team

**Comment by MargotGsb** [comment]
Dear DHBWHub Team, 

I think your selection for the testing metrics is a very good fit for your project. Your achievements for this sprint are also very imprerssive. Especially the OAuth Authorization is a very cool new feature.
We are looking forward to see your further progress.

Kind regards
Your Team Plapy

**Comment by LeonmAndreesen** [comment]
Really good project. We like your weekly report. The progress you made is outstanding.

**Comment by Challenger3010** [comment]
Fantastic overview of software testing metrics! Your detailed breakdown of process, product, and project metrics is very informative and highlights the importance of each type in ensuring software quality. The inclusion of specific examples like defect density and test coverage really helps in understanding how these metrics are applied in practice.

I particularly appreciate the additional focus on coupling and web application performance metrics. Tools like PageSpeed Insights and Google Lighthouse are essential for maintaining a responsive and user-friendly web application.
